File: 2-sem/list2/java/WierszTrojkataPascala.java
package packages;

public class WierszTrojkataPascala {

    int[] tablica;

    public WierszTrojkataPascala(int n) throws ArgumentMniejszyOdZera, ZaDuzyWierszTrojkataPascala {
        if(n < 0)
            throw new ArgumentMniejszyOdZera("Nieprawidlowy numer wiersza");
        tablica = new int[n+1];
        for(int i = 0; i <= n; i++) {
            tablica[i] = wartoscDwumianu(n, i);
            if(tablica[i] < 0)
                throw new ZaDuzyWierszTrojkataPascala("Wartosci wiersza przekraczaja maksymalna wartosc typu liczb calkowitych");
        }
    }
    static int wartoscDwumianu(int n, int k) {
        int wynik = n;
        for(int i=1; i<=k; i++) {
            wynik *= n-i;
            wynik /= i+1;
        }
        return wynik;
    }
    public int wspolczynnik(int m) throws ArgumentWiekszyOdArgumentuKonstruktora, ArgumentMniejszyOdZera {
        if(m < 0)
            throw new ArgumentMniejszyOdZera("liczba spoza zakresu");
        else if(m >= tablica.length)
            throw new ArgumentWiekszyOdArgumentuKonstruktora("liczba spoza zakresu");
        else
            return tablica[m];
    }
}

File: 2-sem/list2/c++/WierszTrojkataPascala.cpp
#include "inc/WierszTrojkataPascala.hpp"
#include "inc/Wyjatki.hpp"
#include "inc/Kontroler.hpp"

WierszTrojkataPascala::WierszTrojkataPascala(std::string n) {
    if(!Kontroler::czyLiczba(n))
        throw new NieLiczba("Nieprawidlowa dana jako numer wiersza");
    else {
        int dlugosc = Kontroler::napisNaLiczbe(n);
        ilosc_elementow = dlugosc + 1;
        if(dlugosc < 0)
            throw new ArgumentMniejszyOdZera("Nieprawidlowy numer wiersza");
        tablica = std::vector<int>(dlugosc + 1);
        for(int i=0; i <= dlugosc; i++) {
            tablica.at(i) = WierszTrojkataPascala::wartoscDwumianu(dlugosc, i);
            if(tablica.at(i) < 0)
                throw new ZaDuzyWierszTrojkataPascala("Wartosc wiersza przekraczaja maksymalna wartosc typu liczb calkowitych");
        }
    }
}

int WierszTrojkataPascala::wartoscDwumianu(int n, int k) {
    if(k == 0)
        return 1;
    int wynik = n;
    for(int i=1; i<k; i++) {
        wynik *= n-i;
        wynik /= i+1;
    }
    return wynik;
}

int WierszTrojkataPascala::wspolczynnik(int m) {
    if(m < 0)
        throw new ArgumentMniejszyOdZera("liczba spoza zakresu");
    else if(m >= ilosc_elementow) 
        throw new ArgumentWiekszyOdArgumentuKonstruktora("liczba spoza zakresu");
    else
        return tablica.at(m);
}